### PR TITLE
implement control from *codeMetadata

### PR DESCRIPTION
**Pull request by paul-r-gall**
Commit Message: implement initial flow control in decode/encodeMetadata
Additional Description: Adds the ability to continue iteration through the filter chain as well as send local replies from both decodeMetadata and encodeMetadata.
Risk Level: Low -- does not change any existing semantics.
Testing: unit testing for FilterManager and integration testing for sending local replies and returning ContinueAll
Docs Changes: 
Release Notes: yes
Platform Specific Features:

### CODE REVIEW

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -777,10 +777,15 @@ void FilterManager::decodeTrailers(ActiveStreamDecoderFilter* filter, RequestTra
 }
 
 void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMap& metadata_map) {
+  ScopeTrackerScopeState scope(&*this, dispatcher_);
+  filter_manager_callbacks_.resetIdleTimer();
```
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:781]: IMO this is worth calling out in release notes as a functional change / bugfix

File: source/common/http/filter_manager.cc [hunk 2]
```diff
@@ -791,12 +796,34 @@ void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMa
       (*entry)->getSavedRequestMetadata()->emplace_back(std::move(metadata_map_ptr));
       return;
     }
-
+    state_.filter_call_state_ |= FilterCallState::DecodeMetadata;
     FilterMetadataStatus status = (*entry)->handle_->decodeMetadata(metadata_map);
-    ASSERT(!hasPreparedLocalReply(), "Sending Local Reply from Metadata is not yet supported.");
+    state_.filter_call_state_ &= ~FilterCallState::DecodeMetadata;
```
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:801]: This doesn't appear parallel to how the other filter state modification works?

**Review by alyssawilk — COMMENTED** [review]
awesome addition.  I'm going to tag Kevin for first pass (with 2 drive by comments)

File: source/common/http/filter_manager.cc (hunk 1 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc:781]: Done.

**Comment by alyssawilk** [comment]
@KBaichoo ping

File: envoy/http/filter.h [hunk 3]
```diff
@@ -174,11 +174,21 @@ enum class FilterTrailersStatus {
 
 /**
  * Return codes for encode metadata filter invocations. Metadata currently can not stop filter
- * iteration.
+ * iteration except in the case of local replies.
  */
 enum class FilterMetadataStatus {
-  // Continue filter chain iteration.
+  // Continue filter chain iteration for metadata only. Does not unblock returns of StopIteration*
+  // from  (decode|encode)(Headers|Data).
```
Comment (KBaichoo) [inline envoy/http/filter.h:181]: nit: two spaces between "from" and "(decode|"

File: source/common/http/filter_manager.cc [hunk 4]
```diff
@@ -1196,9 +1223,29 @@ void FilterManager::encodeMetadata(ActiveStreamEncoderFilter* filter,
...
-                     (*entry)->filter_context_.config_name, static_cast<uint64_t>(status));
+
+    state_.filter_call_state_ &= ~FilterCallState::EncodeMetadata;
+
+    ENVOY_STREAM_LOG(trace, "encode metadata called: filter={} status={} metadata={}", *this,
+                     (*entry)->filter_context_.config_name, static_cast<uint64_t>(status),
+                     *metadata_map_ptr);
+
+    if (status == FilterMetadataStatus::StopIterationForLocalReply) {
+      // The way is shut. It was made by those who are dead, and the dead keep it.
```
Comment (KBaichoo) [inline source/common/http/filter_manager.cc:1237]: Can you illuminate this more?  Why the asymmetry?

Comment (KBaichoo) [reply source/common/http/filter_manager.cc:1237]: Seem like this should also not just fail silently?

File: source/common/http/filter_manager.cc (hunk 1 above)
Comment (KBaichoo) [inline source/common/http/filter_manager.cc:781]: can you add a test now so the fixed bug doesn't regress?

File: test/common/http/filter_manager_test.cc [hunk 5]
```diff
@@ -432,6 +432,123 @@ TEST_F(FilterManagerTest, GetRouteLevelFilterConfigForNullRoute) {
...
+  }
+
+  filter_2->decoder_callbacks_->encodeHeaders(
+      std::make_unique<TestResponseHeaderMapImpl>(*response_headers), false, "none");
+  filter_2->decoder_callbacks_->encodeMetadata(std::make_unique<MetadataMap>(map1));
+
+  filter_manager_->destroyFilters();
+}
+
+TEST_F(FilterManagerTest, MetadataSendsLocalReply) {
```
Comment (KBaichoo) [inline test/common/http/filter_manager_test.cc:508]: can we also test the encoding metadata local reply

File: test/integration/filters/metadata_control_filter.cc [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  }
+  Http::FilterDataStatus encodeData(Buffer::Instance&, bool end_stream) override {
+    return end_stream ? Http::FilterDataStatus::Continue
+                      : Http::FilterDataStatus::StopIterationAndWatermark;
+  }
+  Http::FilterMetadataStatus encodeMetadata(Http::MetadataMap& metadata_map) override {
+    if (metadata_map.contains("should_continue")) {
+      return Http::FilterMetadataStatus::ContinueAll;
+    } else if (metadata_map.contains("local_reply")) {
+      decoder_callbacks_->sendLocalReply(Http::Code::BadRequest, "baaaad", nullptr, absl::nullopt,
```
Comment (KBaichoo) [inline test/integration/filters/metadata_control_filter.cc:49]: should be using the encoder callback here

File: test/integration/protocol_integration_test.cc [hunk 7]
```diff
@@ -2993,6 +2993,180 @@ TEST_P(DownstreamProtocolIntegrationTest, LocalReplyWithMetadata) {
...
+
+  ASSERT_TRUE(upstream_request_->waitForEndStream(*dispatcher_));
+
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, true);
+  ASSERT_TRUE(decoder->waitForEndStream(std::chrono::milliseconds(500)));
+  EXPECT_EQ("200", decoder->headers().getStatusValue());
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_P(ProtocolIntegrationTest, LocalReplyFromEncodeMetadata) {
```
Comment (KBaichoo) [inline test/integration/protocol_integration_test.cc:3075]: wasn't local reply from encode metadata supposed to be blocked as you currently implemented it?

File: test/integration/protocol_integration_test.cc [hunk 8]
```diff
@@ -2993,6 +2993,180 @@ TEST_P(DownstreamProtocolIntegrationTest, LocalReplyWithMetadata) {
...
+    RELEASE_ASSERT(bootstrap.mutable_static_resources()->clusters_size() >= 1, "");
+    ConfigHelper::HttpProtocolOptions protocol_options;
+    protocol_options.mutable_explicit_http_config()
+        ->mutable_http2_protocol_options()
+        ->set_allow_metadata(true);
+    ConfigHelper::setProtocolOptions(*bootstrap.mutable_static_resources()->mutable_clusters(0),
+                                     protocol_options);
+  });
+
+  config_helper_.prependFilter(R"EOF(
```
Comment (KBaichoo) [inline test/integration/protocol_integration_test.cc:3143]: I think it might be helpful to have a test where the succeeding filter does a stopall and we use metadata to restart -- currently these are the same filter

**Review by KBaichoo — COMMENTED** [review]
Thank you for working on this

File: source/common/http/filter_manager.cc (hunk 4 above)
Comment (KBaichoo) [reply source/common/http/filter_manager.cc]: This still fails silently. Can you provide context why we don't want to support this on the encode* route?

File: envoy/http/filter.h (hunk 3 above)
Comment (paul-r-gall) [reply envoy/http/filter.h]: Done.

File: test/integration/filters/metadata_control_filter.cc (hunk 6 above)
Comment (KBaichoo) [reply test/integration/filters/metadata_control_filter.cc]: This should use the encoder callbacks, if we just had an encoder filter it would not have access to decoder callbacks here to do local reply.

File: source/common/http/filter_manager.cc (hunk 1 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc:781]: Done.

File: test/integration/filters/metadata_control_filter.cc (hunk 6 above)
Comment (paul-r-gall) [reply test/integration/filters/metadata_control_filter.cc]: Done.

File: test/integration/protocol_integration_test.cc (hunk 7 above)
Comment (KBaichoo) [reply test/integration/protocol_integration_test.cc:3075]: It's still unclear to me if this is intended given that there's the "path" for local reply in encodeMetadata doesn't seem to contain the same check for local replies.

File: test/integration/protocol_integration_test.cc (hunk 8 above)
Comment (KBaichoo) [reply test/integration/protocol_integration_test.cc:3143]: makes sense, thanks

**Review by KBaichoo — DISMISSED** [review]
Looks good pending CI

File: source/common/http/filter_manager.cc (hunk 4 above)
Comment (KBaichoo) [reply source/common/http/filter_manager.cc]: makes sense

File: envoy/http/filter.h [hunk 9]
```diff
@@ -174,11 +174,20 @@ enum class FilterTrailersStatus {
 
 /**
  * Return codes for encode metadata filter invocations. Metadata currently can not stop filter
- * iteration.
+ * iteration except in the case of local replies.
  */
 enum class FilterMetadataStatus {
-  // Continue filter chain iteration.
+  // Continue filter chain iteration for metadata only. Does not unblock returns of StopIteration*
```
Comment (alyssawilk) [inline envoy/http/filter.h:180]: Is this the behavior for any other filter chain continue? I'd prefer to not add more complexity to the filter chain if we can avoid it.

File: source/common/http/filter_manager.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/filter_manager.cc:801]: I swear I was looking at a version which didn't &=| out state.  This looks fine.

File: source/common/http/filter_manager.h [hunk 10]
```diff
@@ -1012,22 +1012,24 @@ class FilterManager : public ScopeTrackedObject,
     struct FilterCallState {
       static constexpr uint32_t DecodeHeaders   = 0x01;
       static constexpr uint32_t DecodeData      = 0x02;
-      static constexpr uint32_t DecodeTrailers  = 0x04;
-      static constexpr uint32_t EncodeHeaders   = 0x08;
-      static constexpr uint32_t EncodeData      = 0x10;
-      static constexpr uint32_t EncodeTrailers  = 0x20;
+      static constexpr uint32_t DecodeMetadata  = 0x04;
```
Comment (alyssawilk) [inline source/common/http/filter_manager.h:1015]: Why shift these values? I think it was nice they were in order but the requirement really is they're unique.

**Review by alyssawilk — COMMENTED** [review]
If you have time free in my sadly few free slots today let's sync up re continue in person?

File: source/common/http/filter_manager.h (hunk 10 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.h:1015]: I just wanted them in order. No other reason. Happy to revert if you prefer.

File: envoy/http/filter.h (hunk 9 above)
Comment (mattklein123) [reply envoy/http/filter.h:180]: Yeah I agree this is quite confusing. Renaming at minimum sounds fine to me if the alternative is too risky (agreed it probably is)

File: envoy/http/filter.h [hunk 11]
```diff
@@ -174,11 +174,20 @@ enum class FilterTrailersStatus {
...
+  // from (decode|encode)(Headers|Data).
   Continue,
+
+  // Continue filter chain iteration. If headers have not yet been sent to the next filter, they
+  // will be sent first via (decode|encode)Headers().
+  ContinueAll,
+
+  // Stops iteration of the entire filter chain. Only to be used in the case of sending a local
+  // reply from (decode|encode)Metadata.
+  StopIterationForLocalReply,
```
Comment (mattklein123) [inline envoy/http/filter.h:190]: Why is it this specific vs. just StopIteration? Couldn't theoretically someone right a filter in which they stop iteration and then release the metadata (or maybe in the future change it) when they get headers/data?

Comment (paul-r-gall) [reply envoy/http/filter.h:190]: It's this specific because I didn't want to implement FilterMetadataStatus::StopIteration in its full generality yet. If you think it's necessary at this time, I can add it.

Comment (paul-r-gall) [reply envoy/http/filter.h:190]: There's ASSERTs in the code to enforce that this is only returned in cases of local reply...

File: envoy/http/filter.h (hunk 9 above)
Comment (alyssawilk) [reply envoy/http/filter.h:180]: I'd suggested a follow-up, I'd also be fine with you just shipping it out as a separate PR but I'd definitely like to keep the high risk changing-the-filter-chain PR separate from the renaming-a-field-in-many-places-PR :-)
/wait

Comment (paul-r-gall) [reply envoy/http/filter.h:180]: Great -- I reverted the rename, and will move it to a followup PR.

File: test/integration/protocol_integration_test.cc [hunk 12]
```diff
@@ -2997,6 +2997,180 @@ TEST_P(DownstreamProtocolIntegrationTest, LocalReplyWithMetadata) {
   ASSERT_EQ("200", response->headers().getStatusValue());
 }
 
+TEST_P(DownstreamProtocolIntegrationTest, LocalReplyFromDecodeMetadata) {
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3000]: can we move this to filter_integration_test and make sure that for both upstream and downstream if we have
StopAllFilter MetadataLocalReplyFilter ReleaseAssertIfSeesHeadersFilter RouterFilter
that all goes well?
I have some concerns about the filter manager impl and want to make sure tests cover things

**Review by alyssawilk — COMMENTED** [review]
/wait

File: test/integration/protocol_integration_test.cc (hunk 12 above)
Comment (paul-r-gall) [reply test/integration/protocol_integration_test.cc:3000]: Added test that does MetadataLocalReplyFilter -> ReleaseAssertIfSeesHeadersFilter, but am not sure what the purpose of the StopAllFilter is.

File: source/common/http/filter_manager.cc [hunk 13]
```diff
@@ -791,12 +796,36 @@ void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMa
...
-
+    state_.filter_call_state_ |= FilterCallState::DecodeMetadata;
     FilterMetadataStatus status = (*entry)->handle_->decodeMetadata(metadata_map);
-    ASSERT(!hasPreparedLocalReply(), "Sending Local Reply from Metadata is not yet supported.");
+    state_.filter_call_state_ &= ~FilterCallState::DecodeMetadata;
+
     ENVOY_STREAM_LOG(trace, "decode metadata called: filter={} status={}, metadata: {}", *this,
                      (*entry)->filter_context_.config_name, static_cast<uint64_t>(status),
                      metadata_map);
+    if (state_.decoder_filter_chain_aborted_) {
```
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:806]: Ah, so one of the reasons I asked you to e2e test with an abortIFSeesHeader filter is I don't think this works for the upstream filter chain.  decoder_filter_chain_aborted is fundamentally a downstream boolean as it's triggered on local reply.  local replies only hit the downstream filter chain, so I think what will happen if you execute this path on the upstream filter chain is we'll send a local reply but keep iterating on metadata.

Having looked at the filter you used it doesn't crash on metadata.  Can you update that and see if my concern is valid?  If so I'd love another rename TODO for either you or me to rename these booleans  to be more clear 
/wait

Comment (alyssawilk) [reply source/common/http/filter_manager.cc:806]: oh hey look I already safeguarded against this mistake and forgot about it.
https://github.com/envoyproxy/envoy/blob/main/source/common/router/upstream_request.cc#L69

File: test/integration/protocol_integration_test.cc (hunk 12 above)
Comment (alyssawilk) [reply test/integration/protocol_integration_test.cc:3012]: is this now redundant with the filter integration tst?  should we remove?

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo test nits! Throwing over to snow for second pass :-)

File: test/integration/protocol_integration_test.cc [hunk 14]
```diff
@@ -3009,6 +3009,152 @@ TEST_P(DownstreamProtocolIntegrationTest, LocalReplyWithMetadata) {
...
+
+  ASSERT_TRUE(upstream_request_->waitForEndStream(*dispatcher_));
+
+  upstream_request_->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, true);
+  ASSERT_TRUE(decoder->waitForEndStream(std::chrono::milliseconds(500)));
+  EXPECT_EQ("200", decoder->headers().getStatusValue());
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_P(ProtocolIntegrationTest, LocalReplyFromEncodeMetadata) {
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3063]: also redundant?

**Review by alyssawilk — COMMENTED** [review]
one last nit!  I think :-P

**Review by alyssawilk — DISMISSED** [review]

**Comment by alyssawilk** [comment]
/retest

@snowp can you PTAL?

File: source/common/http/filter_manager.cc [hunk 15]
```diff
@@ -791,12 +796,36 @@ void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMa
...
+      // If the decoder filter chain has been aborted, then either:
+      // 1. This filter has sent a local reply from decode metadata.
+      // 2. This filter is the Router filter, and and upstream filter has sent a local reply.
```
Comment (snowp) [inline source/common/http/filter_manager.cc:809]:
```suggestion
      // 2. This filter is the Router filter, and an upstream filter has sent a local reply.
```

File: source/common/http/filter_manager.cc [hunk 16]
```diff
@@ -791,12 +796,36 @@ void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMa
...
+      ASSERT((status == FilterMetadataStatus::StopIterationForLocalReply) ||
+             (std::next(entry) == decoder_filters_.end()));
```
Comment (snowp) [inline source/common/http/filter_manager.cc:811]: Since we're just checking to see that the filter is the last one, maybe reword the second point to talk about this being a "terminal filter" - the router is one of these but the system allows for any terminal filter to be used at the end of the filter chain

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, just one small comment

File: source/common/http/filter_manager.cc (hunk 16 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc:811]: Done, thanks for review!

File: source/common/http/filter_manager.cc (hunk 15 above)
Comment (paul-r-gall) [reply source/common/http/filter_manager.cc]: Done!
